### PR TITLE
VM tracing and JSON RPC endpoint for it.

### PR DESCRIPTION
**Pull request by gavofyork**
An additional type similar in usage to `Tracer` called `VMTracer`. This is to provide all information necessary to reconstruct a full VM trace. Requires a bunch of fairly trivial data structs like `VMTrace`, `VMOperation` &c.

Also fixed `call`s which were using U256's max value and therefore failed when funds were sent to the caller.

mea culpa for the few other minor additions - an extra morden node, support for `Option` in RLP streams and some formatting corrections.

### CODE REVIEW

File: ethcore/src/evm/interpreter.rs [hunk 1]
```diff
@@ -311,10 +324,19 @@ impl evm::Evm for Interpreter {
...
+			};
+
 			// Execute instruction
 			let result = try!(self.exec_instruction(
 					current_gas, &params, ext, instruction, &mut reader, &mut mem, &mut stack
-					));
+			));
+
+			if trace_executed {
+				ext.trace_executed(current_gas, stack.peek_top(get_info(instruction).ret), mem_written.map(|(o, s)| (o, &(mem[o..(o + s)]))), store_written);
```
Comment (tomusdrw) [inline ethcore/src/evm/interpreter.rs]: You could use a closure to get rid of the `if` statement.

**Comment by arkpar** [comment]
Looks good to me

File: ethcore/src/evm/interpreter.rs [hunk 2]
```diff
@@ -293,10 +301,15 @@ impl evm::Evm for Interpreter {
 
 		while reader.position < code.len() {
 			let instruction = code[reader.position];
-			reader.position += 1;
 
 			// Calculate gas cost
 			let (gas_cost, mem_size) = try!(self.get_gas_cost_mem(ext, instruction, &mut mem, &stack));
+
+			// TODO: make compile-time removable if too much of a performance hit.
```
Comment (rphmeier) [inline ethcore/src/evm/interpreter.rs]: the cfg! macro would be useful for that

Comment (gavofyork) [inline ethcore/src/evm/interpreter.rs]: no evidence that it's a bottleneck yet...

File: ethcore/src/evm/interpreter.rs (hunk 1 above)
Comment (gavofyork) [inline ethcore/src/evm/interpreter.rs]: you mean return something implementing a `FnMut`?

Comment (tomusdrw) [inline ethcore/src/evm/interpreter.rs]: `ext.trace_executed(|| (current_gas, stack.peek_top(get_info(instruction).ret), mem_written.map(|(o, s)| (o, &(mem[o..(o + s)]))), store_written))`

you could get rid of the `trace_executed` flag.

Comment (gavofyork) [inline ethcore/src/evm/interpreter.rs]: that's a dynamic dispatched function (via evm::Ext) - probably not worth the performance loss for a single `if`.

**Pull request merged**